### PR TITLE
Single-File: Pass BUNDLE_PROBE property to the runtime

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
As described in the [design doc](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup), pass the bundle_probe function pointer encoded as a string to the runtime.

### CODE REVIEW

File: src/installer/corehost/cli/bundle/runner.cpp [hunk 1]
```diff
@@ -81,3 +81,47 @@ bool runner_t::locate(const pal::string_t& relative_path, pal::string_t& full_pa
...
+
+        assert(*offset != 0);
+
+        return true;
+    }
+
+    return false;
+}
+
+pal::string_t runner_t::get_bundle_probe()
```
Comment (vitek-karas) [inline src/installer/corehost/cli/bundle/runner.cpp:116]: Just my personal preference: I would move this to hostpolicy coreclr setup code. I don't think the runner is the right component to know about who we encode the probe value to pass it to the runtime - that's a concern which is solved by hostpolicy coreclr interaction code.

File: src/installer/corehost/cli/bundle/runner.h [hunk 2]
```diff
@@ -37,8 +37,22 @@ namespace bundle
...
+        // Probe the app-bundle for the file 'path' and return its location ('offset', 'size') if found.
+        // This method is intnded to be used by the runtime to probe for bundled assemblies
+        // This method assumes that the currently executing app is a single-file bundle.
+        //
+        // bundle_probe recieves its path argument as wchar_t* instead of pal::char_t*, because:
+        // * The host uses Unicode strings on Windows and UTF8 strings on Unix
+        // * The runtime uses Unicode strings on all platforms
+        // Using a unicode encoded path presents a uniform interface to the runtime
+        // and minimizes the number if Unicode <-> UTF8 conversions necessary.
+        static bool __stdcall bundle_probe(const wchar_t* path, int64_t* offset, int64_t* size);
```
Comment (vitek-karas) [inline src/installer/corehost/cli/bundle/runner.h:54]: I would make this a top-level/global function and not a member function - to make it clear it's effectively an exported API. Or maybe even better - do the same which hostpolicy already does for functions like `get_delegate` and similar. Those are also effectively exported APIs which are returned as function pointers. Currently they are written as hidden inside an anonymous namespace and don't need header/cpp duplicate declarations.

Also currently the header and cpp have slightly different declarations of the function - which feels wrong.

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp [hunk 3]
```diff
@@ -13,9 +13,12 @@
 #include <fcntl.h>
 #include <fnmatch.h>
 #include <ctime>
+#include <clocale>
```
Comment (vitek-karas) [inline src/installer/corehost/cli/hostmisc/pal.unix.cpp:16]: Can you please make sure that this change doesn't regress the size of apphosts? In theory it should not as the new functionality is not used by apphosts, but still.

**Review by vitek-karas — APPROVED** [review]
Just minor stuff, looks good.

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp (hunk 3 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/hostmisc/pal.unix.cpp:16]: On Linux X64 Release build, apphost binary included the `unicode_palstring` function -- so it wasn't removed by the compiler. But the apphost size only increased by 16 bytes (the function's size is definitely bigger than 16 bytes).

**Review by vitek-karas — APPROVED** [review]

**Comment by swaroop-sridhar** [comment]
@vitek-karas  I updated the `bundle_probe` interface from 
```
bool STDMETHODCALLTYPE bundle_probe(const wchar_t* path, int64_t* offset, int64_t* size)
```
to 
```
bool STDMETHODCALLTYPE bundle_probe(const char16_t* path, int64_t* offset, int64_t* size)
```
because:

* `wchar_t` is 16-bit encoding on Windows while it is 32-bit encoding on most Unix systems
* The runtime uses 16-bit encoded unicode characters.

Therefore, the interface uses the fixed char16_t type. This fixed the test failures on Linux.

**Pull request merged**